stdlib/public/core/StringNormalization.swift

```diff
@@ -102,3 +112,331 @@ internal func _tryNormalize(
+private func transcodeSegmentToUTF16(
+  _ sourceBuffer: UnsafeBufferPointer<UInt8>,
+  into outputBuffer: UnsafeMutableBufferPointer<UInt16>
+) -> (Int, Int)? {
+  var readIndex = 0
+  var writeIndex = 0
+  let outputCount = outputBuffer.count
+  let sourceCount = sourceBuffer.count
+  
+  while readIndex < sourceCount {
+    let (scalar, length) = _decodeScalar(sourceBuffer, startingAt: readIndex)
+    
+    if scalar._hasNormalizationBoundaryBefore && readIndex != 0 {
+      break
+    }
+    
+    readIndex += length
+    
+    for cu in scalar.utf16 {
+      if writeIndex < outputCount {
+        outputBuffer[writeIndex] = cu
+        writeIndex += 1
+      } else {
+        return nil
+      }
+    }
+  }
+  
+  return (readIndex, writeIndex)
+}
+
+//transcodes the UTF16 segment stored in soureceBuffer into the outputBuffer as UTF8
+private func transcodeSegmentToUTF8(
+  _ sourceBuffer: UnsafeBufferPointer<UInt16>,
+  into outputBuffer: UnsafeMutableBufferPointer<UInt8>
+) -> (Int, Int)? {
+  var readIndex = 0
+  var writeIndex = 0
+  let outputCount = outputBuffer.count
+  let sourceCount = sourceBuffer.count
+  
+  while readIndex < sourceCount {
+    let (scalar, length) = _decodeScalar(sourceBuffer, startingAt: readIndex)
+    //we don't need to check for normalization boundaries here because we are only transcoding
+    //a single segment at this point
+    
+    readIndex += length
+    
+    for cu in UTF8.encode(scalar)._unsafelyUnwrappedUnchecked {
+      if writeIndex < outputCount {
+        outputBuffer[writeIndex] = cu
+        writeIndex &+= 1
+      } else {
+        return nil
+      }
+    }
+  }
+  return (readIndex, writeIndex)
+}
+
+internal func _reallocateBuffers(
+  sourceCount count: Int,
+  outputBuffer: inout UnsafeMutableBufferPointer<UInt8>,
+  icuInputBuffer: inout UnsafeMutableBufferPointer<UInt16>,
+  icuOutputBuffer: inout UnsafeMutableBufferPointer<UInt16>
+) {
+  let output = count * _Normalization._maxNFCExpansionFactor * _Normalization._maxUTF16toUTF8ExpansionFactor
+  let icuInput = count
+  let icuOutput = count * _Normalization._maxNFCExpansionFactor
+  outputBuffer = UnsafeMutableBufferPointer<UInt8>.allocate(capacity: output)
+  icuInputBuffer = UnsafeMutableBufferPointer<UInt16>.allocate(capacity: icuInput)
+  icuOutputBuffer = UnsafeMutableBufferPointer<UInt16>.allocate(capacity: icuOutput)
+}
+
+internal func _fastNormalize(
+  readIndex: String.Index,
+  sourceBuffer: UnsafeBufferPointer<UInt8>,
+  outputBuffer: inout UnsafeMutableBufferPointer<UInt8>,
+  icuInputBuffer: inout UnsafeMutableBufferPointer<UInt16>,
+  icuOutputBuffer: inout UnsafeMutableBufferPointer<UInt16>
+) -> NormalizationResult {
+  let start = readIndex.encodedOffset
+  let rebasedSourceBuffer = UnsafeBufferPointer(rebasing: sourceBuffer[start...])
+  do {
+    let (read, filled) = fastFill(rebasedSourceBuffer, outputBuffer)
+      if filled > 0 {
+        let nextIndex = readIndex.encoded(offsetBy: read)
```

Comment:
We can have `fastFill` return optional similarly to transcode. Then, no need for `do` block etc., they're all if/guard lets